network/p2p/tracer/internal/duplicate_msgs_counter_cache.go

```diff
@@ -0,0 +1,166 @@
+		// perform recordDecay on gauge value
+		record := duplicateMessagesCounter(entity)
+		entity, err = g.decayAdjustment(record)
+		return entity
+	})
+	if err != nil {
+		return 0, fmt.Errorf("unexpected error while applying recordDecay adjustment for peerID %s: %w", peerId, err)
+	}
+	if !adjusted {
+		return 0, fmt.Errorf("unexpected error record not found for  peerID %s, even after an init attempt", peerId)
+	}
+
+	record := duplicateMessagesCounter(adjustedEntity)
+	return record.Gauge, nil
+}
+
+// Inc increments the number of duplicate messages detected for the peer. This func is used in conjunction with the GossipSubMeshTracer and is invoked
+// each time the DuplicateMessage callback is invoked.
+// Args:
+// - peerID: the peer ID of the peer in the GossipSub protocol.
+//
+// Returns:
+// - float64: updated value for the duplicate message tracker.
+// - error: if any error was encountered during record adjustments in cache.
+// No errors are expected during normal operation.
+func (g *GossipSubDuplicateMessageTrackerCache) Inc(peerId peer.ID) (float64, error) {
+	var err error
+	entityID := flow.MakeID(peerId)
+	optimisticAdjustFunc := func() (flow.Entity, bool) {
+		return g.c.Adjust(entityID, func(entity flow.Entity) flow.Entity {
+			record := duplicateMessagesCounter(entity)
+			entity, err = g.decayAdjustment(record) // first recordDecay the record
+			if err != nil {
+				return entity
+			}
+			return g.incrementAdjustment(entity) // then increment the record
+		})
+	}
+
+	// optimisticAdjustFunc is called assuming the record exists; if the record does not exist,
+	// it means the record was not initialized. In this case, initialize the record and call optimisticAdjustFunc again.
+	// If the record was initialized, optimisticAdjustFunc will be called only once.
+	adjustedEntity, adjusted := optimisticAdjustFunc()
+	if err != nil {
+		return 0, fmt.Errorf("unexpected error while applying recordDecay adjustment for peerID %s: %w", peerId, err)
+	}
+	if !adjusted {
+		g.init(peerId)
+		adjustedEntity, adjusted = optimisticAdjustFunc()
+		if !adjusted {
+			return 0, fmt.Errorf("unexpected record not found for peerID %s, even after an init attempt", peerId)
+		}
+	}
+
+	return adjustedEntity.(DuplicateMessagesCounterEntity).Gauge, nil
+}
+
+// incrementAdjustment performs a cache adjustment that increments the guage for the DuplicateMessagesCounterEntity
+func (g *GossipSubDuplicateMessageTrackerCache) incrementAdjustment(entity flow.Entity) flow.Entity {
+	record := duplicateMessagesCounter(entity)
+	record.Gauge++
+	record.lastUpdated = time.Now()
+	// Return the adjusted record.
+	return record
+}
+
+// decayAdjustment performs geometric recordDecay on the duplicate message counter gauge of a peer. This ensures a peer is not penalized forever.
+func (g *GossipSubDuplicateMessageTrackerCache) decayAdjustment(counter DuplicateMessagesCounterEntity) (DuplicateMessagesCounterEntity, error) {
+	received := counter.Gauge
```

Comment:
```suggestion
	received := counter.Value
```
counter.Gauge may seem a bit confusing.

Reply:
https://github.com/onflow/flow-go/pull/5058/commits/57fc14ab8d0e531bc517a2d09e6373fd51b13584